File: 0042-Trapping-Rain-Water/lc0042.js
// JavaScript solution.

/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function(height) {
    let left = 0, right = height.length-1;
    let leftmax = 0, rightmax = 0;
    let ans = 0;
    while (left < right) {
        if (height[left] < height[right]) {
            leftmax = Math.max(leftmax, height[left]);
            ans += leftmax - height[left];
            left++;
        } else {
            rightmax = Math.max(rightmax, height[right]);
            ans += rightmax - height[right];
            right--;
        }
    }
    return ans;
};

File: 0042-Trapping-Rain-Water/lc0042.py
# Python solution.

class Solution:
    def trap(self, height: List[int]) -> int:
        left, right = 0, len(height)-1
        leftmax, rightmax = 0, 0
        ans = 0
        while left < right:
            if height[left] < height[right]:
                leftmax = max(leftmax, height[left])
                ans += leftmax - height[left]
                left += 1
            else:
                rightmax = max(rightmax, height[right])
                ans += rightmax - height[right]
                right -= 1
        return ans
